### ISSUE TITLE
touch: cannot touch ‘/var/jenkins_home/copy_reference_file.log’: Permission denied

### ISSUE THREAD

**Issue by mmaryo** [jenkinsci/docker#493]
I have the same issue like this : https://github.com/jenkinsci/docker/issues/177

`$ docker run -ti -p 8080:8080 -p 50000:50000 -v /opt/jenkins:/var/jenkins_home jenkins 
touch: cannot touch ‘/var/jenkins_home/copy_reference_file.log’: Permission denied
Can not write to /var/jenkins_home/copy_reference_file.log. Wrong volume permissions?
`

Any idea ?

**Comment by carlossg** [comment]
as mentioned there you need to figure out your volume mapping permissions 
ie.

    sudo chown -R 1000:1000 /opt/jenkins

**Comment by mmaryo** [comment]
it works thanks :)

**Comment by egbertzeng** [comment]
it works for me ,  thanks!

**Comment by Padhh** [comment]
works for me, thanks

**Comment by harigitgub** [comment]
docker run -it --rm --name jenkins -p 8090:8080 -p 50 000:50000 -v /home/docker/jenkins:/var/jenkins_home csanchez/jenkins-kubernetes

after running this command and checking in the browser with http://ip:8090 it is asking for password after installing of jenkins.
As password is shown in terminal i entered and it asking to install plugins in which kubernetes plugin not available and  screen stuck there not moving forward

any suggestions on this

**Comment by deepforu47** [comment]
I am still getting permission issues-
# ll -nd /home/jenkins/jenkin_home/
drwxrwxr-x. 2 1000 1000 6 Dec  8 15:57 /home/jenkins/jenkin_home/

~]# docker run -it  --rm -p 8080:8080 -p 50000:50000 -v /home/jenkins/jenkin_home:/var/jenkins_home jenkins/jenkins:lts
touch: cannot touch '/var/jenkins_home/copy_reference_file.log': Permission denied
Can not write to /var/jenkins_home/copy_reference_file.log. Wrong volume permissions?

**Comment by fasolens** [comment]
What is your UID of jenkins user inside docker container? Is it also 1000?

**Comment by junxinzhang** [comment]
You solved my problem, very good, thank you

**Comment by batmat** [comment]
Side note for future people passing by: the original question/issue also comes from a usage that is generally strongly discouraged, i.e. using the often so-called _bind mounts_.

More precisions are given on https://docs.docker.com/storage/bind-mounts/, but basically using a mapping from the container to the host often leads (as you noticed) to all variations of permissions issues. 

The strong recommendation, by the Docker community in general, and also by us on Jenkins side, is to use a volume instead.

More concretely, the command to run can simply be adjusted to:

```shell
docker run -ti -p 8080:8080 -p 50000:50000 -v jenkins-data:/var/jenkins_home jenkins 
```

This command will create a named volume called `jenkins-data` where your Jenkins home will be stored. And you will not hit those weird permissions anymore.

```shell
docker volume ls
DRIVER              VOLUME NAME
local               jenkins-data
```

Showing how to access the data to do whatever you want, check or run a backup:

Example mounting the volume in read-only:

```shell
docker run -ti -v jenkins-data:/var/jenkins_home:ro debian bash
root@9e8f2c20ca2a:/# ls /var/jenkins_home/
config.xml                     jenkins.CLI.xml                      nodes                     updates
copy_reference_file.log        jenkins.install.UpgradeWizard.state  plugins                   userContent
hudson.model.UpdateCenter.xml  jobs                                 secret.key                users
identity.key.enc               logs                                 secret.key.not-so-secret  war
init.groovy.d                  nodeMonitors.xml                     secrets
root@9e8f2c20ca2a:/#
```

**Comment by jcrben** [comment]
@batmat bind mounts are typically used to ease development since changes are synced, and I'm not sure that volumes help in that regard unless you're somehow getting inside the container or volume or using nfs or something...?

**Comment by batmat** [comment]
@jcrben I strongly doubt people here are all *developing* on Jenkins. I can totally be wrong, but I would think people are just trying to get a Jenkins server up running as a Docker container. Hence these people should just not look for anything else than a volume.

Bind mounts _can_ be used to do things, but though they look easy at first sight, they are not. Using bind mounts correctly requires a very clear understanding of the Docker and Linux permission model, so that one can pass the right `--user`, or change permissions on the host (urgh, do not do that IMO).

In other words, if you do know exactly what those caveats are, you would not land here asking for a (quick) solution.

**The only reasonable solution for non-Docker experts is to use volumes**, is what I'm saying. Using bind-mounts without a deep understanding of those is going to be *very* painful.

**Comment by jcrben** [comment]
I'd be interested in reading material on all these best practices if you have one handy. I stumbled thru this and I'm not entirely sure how I fixed it. My disorganized notes are at https://gitlab.com/jcrben-staples/knowledge/blob/master/programming/devops_scaling/docker_permissions_mac.md

While I'm not developing on jenkins, I use bind mounts to ease the automation setup - for example, with https://github.com/jenkinsci/configuration-as-code-plugin I can bind mount `jenkins.yaml` and there's a button to reload it.

**Comment by s530665** [comment]
why we need to change ownership permissions

**Comment by pdsuwwz** [comment]
> as mentioned there you need to figure out your volume mapping permissions
> ie.
> 
> ```
> sudo chown -R 1000:1000 /opt/jenkins
> ```

good!

**Comment by scndel** [comment]
To fix easily this permission issue with volume in Kubernetes : just define `jenkins_home` as **subpath** !
```
volumeMounts:
        - mountPath: /var
          name: jenkins-volume
          subPath: jenkins_home
```
My _jenkins_ pod now works fine with Rancher & Longhorn. As default storage class, it has provisoned automatically the volume and I didn't have to write it myself.
<img width="765" alt="jenkins volume kubernetes" src="https://user-images.githubusercontent.com/17161227/79554138-f06b3500-809d-11ea-8d3e-aaff6a479a16.png">

**Comment by micuffaro** [comment]
@scndel , Thank you very much for your input. 
This has solved my issue! Have a great day.

**Comment by guru905** [comment]
> works for me, thanks

Hi am getting below error while installing the Jenkins on docker. Please give me suggestions how i can i resolve this issue.
 
touch: cannot touch ‘/var/jenkins_home/copy_reference_file.log’: Permission denied

Thanks in adavance.

**Comment by VeeraNadikatla** [comment]
@guru905
1. Ensure that the id of the user-id on the host and inside the docker-container are the same.
2. If the above one is clear, check the SELinux status, which can be a road block.

**Comment by skyhitnow** [comment]
hi dudes.I also got that problem. The cause was I powered off the host abruptly,then after I started the server again I found that issue emerged. I finally deleted the copyxxxx.log file,as I thought it's irrelevant as just a log. Then the problem seemly was resolved.FYI.

**Comment by mithlajkn** [comment]
Add this under spec. it may work

initContainers:
      - name: fix-permissions
        image: busybox
        command: ["sh", "-c", "chown -R 1000:1000 /var/jenkins_home"]
        securityContext:
          privileged: true
😃

**Comment by harshblog150** [comment]
> 


> > as mentioned there you need to figure out your volume mapping permissions
> > ie.
> > ```
> > sudo chown -R 1000:1000 /opt/jenkins
> > ```
> 
> good!

@pdsuwwz No Luck!!

getting similar issue - i was trying mount my local-machine All jenkins configurations (including plugins, jobs, workspace) everything into a new jenkins container... No Luck!

[jenkins@ip-10-0-1-12 jenkins]$ chown -R 1001:1001 /home/jenkins/jenkins_home/jenkins

[jenkins@ip-10-0-1-12 jenkins]$ docker run -dit --name Jenkins_Container -p 8084:8080 -p 50000:50000 -v /home/jenkins/jenkins_home/jenkins:/var/jenkins_home jenkins/jenkins:lts-jdk11
57b867bc7a442f5b326153c9454bb8d4dc4989174338c5321d93f0b24ba0da1e
[jenkins@ip-10-0-1-12 jenkins]$ docker ps -a
CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
57b867bc7a44 jenkins/jenkins:lts-jdk11 "/sbin/tini -- /usr/…" 4 seconds ago Exited (1) 3 seconds ago

[jenkins@ip-10-0-1-12 jenkins]$ docker logs -f 57b867bc7a44
touch: cannot touch '/var/jenkins_home/copy_reference_file.log': Permission denied
Can not write to /var/jenkins_home/copy_reference_file.log. Wrong volume permissions?

**Comment by majorgear** [comment]
> Add this under spec. it may work for kubernets
> 
> ```
> initContainers:
>       - name: fix-permissions
>         image: busybox
>         command: ["sh", "-c", "chown -R 1000:1000 /var/jenkins_home"]
>         securityContext:
>           privileged: true
>         volumeMounts:
>         - name: jenkins-vol
>           mountPath: /var/jenkins_home
> ```
> 
> 😃

This fixed it for me! I ran into the issue while migrating Jenkins from docker compose into my k3s cluster.  Thanks!

**Comment by cstichlberger** [comment]
In my case (rootless podman with /var/jenkins_home in a bind mount), using the

`--userns=keep-id`

option with podman run also solved the issue (which most likely only works due to my out-of-container user and in-container jenkins having both the uid 1000)

**Comment by salazar-codes** [comment]
> as mentioned there you need to figure out your volume mapping permissions ie.
> 
> ```
> sudo chown -R 1000:1000 /opt/jenkins
> ```

Works for me! I just had to change the route to my own directory path
